File: android/core/src/main/java/com/vachak/engine/mock/MockEngines.kt
package com.vachak.engine.mock

import com.vachak.engine.*

/**
 * DEV FIXTURES — mock engine implementations used by unit tests and as the
 * default offline build stand-in until real models are wired. They produce
 * deterministic, clearly-fake output and MUST NOT be mistaken for real
 * inference. Swap these for sherpa-onnx / IndicTrans2 adapters in production
 * via EngineProvider without changing any UI code.
 */
object MockTranslationEngine : TranslationEngine {
    override fun supports(pair: LanguagePair) = pair.source == "hi" && pair.target == "mund"
    override fun loadModel(packId: String) = EngineResult.Ok(Unit)
    override fun translate(text: String, pair: LanguagePair): EngineResult<String> =
        if (supports(pair)) EngineResult.Ok("[DEV-FIXTURE-mund] $text")
        else EngineResult.Err(EngineError.UNSUPPORTED_LANGUAGE, "pair $pair unsupported by mock")
}

object MockAsrEngine : ASREngine {
    override fun supports(language: String) = language == "hi"
    override fun loadModel(packId: String) = EngineResult.Ok(Unit)
    override fun transcribe(pcm16: ShortArray, sampleRateHz: Int): EngineResult<String> {
        require(sampleRateHz > 0) { "sampleRateHz must be positive" }
        return if (pcm16.isEmpty()) EngineResult.Err(EngineError.INVALID_INPUT, "empty pcm")
        else EngineResult.Ok("[DEV-FIXTURE-asr] नमस्ते")
    }
}

object MockTtsEngine : TTSEngine {
    override fun supports(language: String) = language == "mund"
    override fun loadModel(packId: String) = EngineResult.Ok(Unit)
    override fun synthesize(text: String, language: String): EngineResult<ShortArray> {
        if (!supports(language)) return EngineResult.Err(EngineError.UNSUPPORTED_LANGUAGE, language)
        // deterministic silence-ish pcm, length scales with text (placeholder)
        return EngineResult.Ok(ShortArray(text.length) { (it % 2).toShort() })
    }
}

object MockCurriculumEngine : CurriculumEngine {
    override fun listLessons(grade: Int) =
        EngineResult.Ok(listOf(LessonRef("L1", "DEV-FIXTURE lesson $grade", grade)))
    override fun getLesson(id: String) = EngineResult.Ok(
        Lesson(id, "DEV-FIXTURE", 1, "हिन्दी पाठ", "[DEV-FIXTURE-mund] lesson", precomputed = true)
    )
    override fun getOutcomes(lessonId: String) =
        EngineResult.Ok(listOf(Outcome("O1", "DEV-FIXTURE outcome", nipunMapped = true)))
}

object MockWorksheetEngine : WorksheetEngine {
    override fun generate(lessonId: String, template: String) = EngineResult.Ok(
        Worksheet("W_$lessonId", lessonId, template,
            listOf(WorksheetItem("DEV-FIXTURE prompt", "DEV-FIXTURE key")))
    )
}

object MockFlashcardEngine : FlashcardEngine {
    override fun listDeck(lessonId: String) = EngineResult.Ok(
        FlashcardDeck(lessonId, listOf(Flashcard("हाथी", "[DEV-FIXTURE] elephant", "asset://elephant.png")))
    )
}

File: shared/schemas/mock_engines.py
"""DEV FIXTURE mock engines for the Python side. Mirror the Kotlin mocks in
android/app/.../engine/mock/MockEngines.kt. Output is clearly fake and must not
be treated as real inference. Swap for real adapters via the same injection
point without changing caller code.
"""
from __future__ import annotations

from .engines import (
    ASREngine, BenchmarkReport, BenchmarkRunner, CurriculumEngine, EngineError,
    Err, Flashcard, FlashcardDeck, FlashcardEngine, LanguagePair, LanguagePackManager,
    Lesson, LessonRef, Ok, Outcome, PackInfo, SyncManager, SyncReport, TTSEngine,
    TranslationEngine, Worksheet, WorksheetEngine, WorksheetItem,
)


class MockTranslationEngine:
    def supports(self, pair: LanguagePair) -> bool:
        return pair.source == "hi" and pair.target == "mund"
    def load_model(self, pack_id: str):
        return Ok(None)
    def translate(self, text: str, pair: LanguagePair):
        if not self.supports(pair):
            return Err(EngineError.UNSUPPORTED_LANGUAGE, f"pair {pair} unsupported")
        return Ok(f"[DEV-FIXTURE-mund] {text}")


class MockASREngine:
    def supports(self, language: str) -> bool:
        return language == "hi"
    def load_model(self, pack_id: str):
        return Ok(None)
    def transcribe(self, pcm16: bytes, sample_rate_hz: int):
        if sample_rate_hz <= 0:
            return Err(EngineError.INVALID_INPUT, "sample_rate_hz must be > 0")
        if not pcm16:
            return Err(EngineError.INVALID_INPUT, "empty pcm")
        return Ok("[DEV-FIXTURE-asr] नमस्ते")


class MockTTSEngine:
    def supports(self, language: str) -> bool:
        return language == "mund"
    def load_model(self, pack_id: str):
        return Ok(None)
    def synthesize(self, text: str, language: str):
        if not self.supports(language):
            return Err(EngineError.UNSUPPORTED_LANGUAGE, language)
        return Ok(b"\x00\x01" * len(text))  # placeholder pcm bytes


class MockCurriculumEngine:
    def list_lessons(self, grade: int):
        return Ok([LessonRef("L1", f"DEV-FIXTURE lesson {grade}", grade)])
    def get_lesson(self, id: str):
        return Ok(Lesson(id, "DEV-FIXTURE", 1, "हिन्दी पाठ", "[DEV-FIXTURE-mund] lesson"))
    def get_outcomes(self, lesson_id: str):
        return Ok([Outcome("O1", "DEV-FIXTURE outcome", True)])


class MockWorksheetEngine:
    def generate(self, lesson_id: str, template: str):
        return Ok(Worksheet(f"W_{lesson_id}", lesson_id, template,
                            [WorksheetItem("DEV-FIXTURE prompt", "DEV-FIXTURE key")]))


class MockFlashcardEngine:
    def list_deck(self, lesson_id: str):
        return Ok(FlashcardDeck(lesson_id, [Flashcard("हाथी", "[DEV-FIXTURE] elephant", "asset://elephant.png")]))


class MockLanguagePackManager:
    def __init__(self):
        self._store: dict[str, PackInfo] = {}
    def install(self, pack_path: str):
        info = PackInfo(pack_path, "mund", "0.1.0", 28, 0)
        self._store[info.id] = info
        return Ok(info)
    def uninstall(self, pack_id: str):
        if self._store.pop(pack_id, None) is None:
            return Err(EngineError.IO_ERROR, "not installed")
        return Ok(None)
    def installed(self):
        return Ok(list(self._store.values()))
    def validate(self, pack_id: str):
        return Ok(pack_id in self._store)
    def rollback(self, pack_id: str):
        if pack_id not in self._store:
            return Err(EngineError.IO_ERROR, "no prior version")
        return Ok(self._store[pack_id])
    def storage_used_bytes(self):
        return Ok(sum(i.size_bytes for i in self._store.values()))


class MockSyncManager:
    def is_network_allowed(self) -> bool:
        return False
    def install_package(self, path: str):
        return Ok(SyncReport(path, "0.1.0", 0, True))
    def last_sync(self):
        return Ok(None)


class MockBenchmarkRunner:
    def run(self, pair: LanguagePair, sample: str):
        r = BenchmarkReport(200, 120, 300, 620, True)
        return Ok(r)
